tempesta_fw/tls.c

```diff
@@ -19,73 +19,583 @@
  * this program; if not, write to the Free Software Foundation, Inc., 59
  * Temple Place - Suite 330, Boston, MA 02111-1307, USA.
  */
+#include "cfg.h"
 #include "connection.h"
+#include "client.h"
 #include "tls.h"
 
+#define tls_dbg(c, fmt, ...)					\
+	TFW_DBG("TLS(%p,%p):%s " fmt "\n", c, c->sk, __func__, ## __VA_ARGS__)
+#define tls_err(c, fmt, ...)					\
+	TFW_ERR("TLS(%p,%p):%s " fmt "\n", c, c->sk, __func__, ## __VA_ARGS__)
+
+#define MAX_TLS_PAYLOAD		1460
+
+typedef struct {
+	mbedtls_ssl_config	cfg;
+	mbedtls_x509_crt	crt;
+	mbedtls_pk_context	key;
+} TfwTls;
+
+static TfwTls tfw_tls = { 0 };
+
+/* Cache for fast TfwTlsContext{} allocations. */
+static struct kmem_cache *tfw_tls_cache = NULL;
+
+static inline int
+ttls_ssl_handshake(TfwTlsContext *tls)
+{
+	int r;
+
+	spin_lock(&tls->lock);
+	r = mbedtls_ssl_handshake(&tls->ssl);
+	spin_unlock(&tls->lock);
```

Comment:
Bad point. We introduced work queue to make sockets CPU local. So that's bad to access TLS context from many CPUs and we should use similar mechanism to ensure that TLS context are per-CPU. Probably not in this patch set, but it's required to finish #81. So please write TODO comment here.